### DISCUSSION TITLE
Indicate field is nullable in swagger/openapi

### DISCUSSION THREAD

**Discussion by jurasofish** [fastapi/fastapi#6910] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI
from pydantic import BaseModel, Field
from typing import Optional

app = FastAPI()


class MyResponseAmbiguous(BaseModel):
    a: Optional[int]

@app.get("/ambiguous", response_model=MyResponseAmbiguous)
def ambiguous():
    # swagger shows ``a`` as type int, even though it can be None
    return MyResponseAmbiguous(a=1)



class MyResponseNullable(BaseModel):
    a: Optional[int] = Field(..., nullable=True)

@app.get("/nullable", response_model=MyResponseNullable)
def nullable():
    # swagger shows ``a`` as type int, and also says it can be nullable.
    return MyResponseNullable(a=1)

```

### Description

When generating an openapi schema, FastAPI does not mark ``Optional`` items as nullable - as in the ``MyResponseAmbiguous`` model above.
To mark an item as nullable in openapi you need to manually pass ``nullable=True`` to the pydantic Field function - as in the ``MyResponseNullable`` model above.
How can this be done automatically?

Additionally, one of the models has ``a`` marked as required (red asterisk) and the other does not - is this the intended behaviour? I think it should be marked as required in both models.

<img width="401" alt="Screen Shot 2021-03-10 at 4 45 25 pm" src="https://user-images.githubusercontent.com/33386122/110582238-05721a80-81c0-11eb-9afb-6b9e62265a2e.png">

### Environment

* OS: Macos
* FastAPI Version: 0.63.0
* Python version: 3.7

**Comment by Mause** [comment]
This is a duplicate of https://github.com/tiangolo/fastapi/issues/2510

**Comment by jurasofish** [comment]
Thanks!

**Comment by gjcarneiro** [comment]
> This is a duplicate of #2510

IMHO it is not a dup of that bug, it's merely related.

* #2510 is: "field with nullable=True doesn't reflect that in the openapi schema/docs;
* this bug is "can we please infer that nullable=True if the field python type is Optional[something]"

We just had a bunch of lost time in my team due to this bug:
* `myfield Optional[int]` correctly causes the field to be marked as optional in the docs
* but I changed to `parent_group_id: Optional[int] = Field(...)` in order to ensure that the field is always present, even if null; but now the docs don't reflect that it is optional.  Unfortunately I missed that the docs changed this way.

As a result, a bunch of time wasted back and forth between frontend and backend devs, because the documentation didn't clearly reflect which fields were optional.

I shame, really.  FastAPI is amazing at time saving and automatically generating docs.  Would be so much nicer if it automatically deduced this one as well.

**Reply by ybressler** [reply]
I understand the frustration here. It's important to point out that annotating with `Optional[T]` results in `Union[T, None]` -> however, does not set the `default value to None`. (This is true for dataclasses, and Pydantic V2.)
```
from typing import Optional
from pydantic import BaseModel


class Model(BaseModel):
    a: Optional[int]


# The following will succeed
my_model = Model(a=None)

# The following will fail
my_model = Model()
```

Alternatively, setting a default value results in a field being truly optional:
```
class Model(BaseModel):
    a: Optional[int] = None


# The following will succeed
my_model = Model(a=None)
my_model = Model()
```

Pydantic's `Field` function will result in the same outcome. I'm not familiar if `FastApi` wraps this function or not – if yes, this is an opportunity to deviate from Pydantic's behavior, though I'm not sure you will want to. Setting `nullable=True` probably is an OpenAPi spec. To achieve this behavior truly, you will need to set default value as `None`.

Perhaps this is where work can be done -> if a field is marked as `nullable=True` and no default is provided, set the default value to `None`.

**Comment by jlplenio** [comment]
> * but I changed to `parent_group_id: Optional[int] = Field(...)` in order to ensure that the field is always present, even if null; but now the docs don't reflect that it is optional.  Unfortunately I missed that the docs changed this way.

With this setup, did you add the nullable=True flag, to indicated in the docs, that the required field is nullable?

**Comment by gjcarneiro** [comment]
Yeah, from reading these comments, I figured adding `nullable=True` would solve my problem.  And it did.  

I just wish things were a bit more automatic.  IMHO, with `Optional[int] = Field(...)`, the Optional should imply that the field is nullable and required, by default.  Instead, by default it implies that the field is required and not nullable.

**Reply by ybressler** [reply]
Updated the docs in Pydantic to reflect this: https://github.com/pydantic/pydantic/pull/6468

Perhaps we should update the docs in FastAPI too? (I'll open an issue)

**Comment by jlplenio** [comment]
> I just wish things were a bit more automatic. IMHO, with `Optional[int] = Field(...)`, the Optional should imply that the field is nullable and required, by default. Instead, by default it implies that the field is required and not nullable.

Thank you for your swift response. I agree with your statement, entirely.

**Comment by fzyzcjy** [comment]
Hi, how can I see `nullable` in swagger/openapi using the following code:

```
class A(BaseModel):
    a: Optional[int]
```

Thanks!

Or, do I have to write down `Field(nullable=True)` every time? IMHO fastapi says it does not repeat itself, so writing `Optional` and `nullable=True` seems to be a duplication.

**Reply by anandtripathi5** [reply]
Tagging @tiangolo here for more visibility as for bigger projects this duplication is causing a lot of overhead

**Comment by andrew222651** [comment]
This is arguably a bug. But the root issue is in Pydantic, see https://github.com/samuelcolvin/pydantic/issues/1270.

**Comment by ivantichy** [comment]
> Hi, how can I see `nullable` in swagger/openapi using the following code:
> 
> ```
> class A(BaseModel):
>     a: Optional[int]
> ```
> 
> Thanks!
> 
> Or, do I have to write down `Field(nullable=True)` every time? IMHO fastapi says it does not repeat itself, so writing `Optional` and `nullable=True` seems to be a duplication.

IMHO "required" and "nullable" are two different things.
https://stackoverflow.com/questions/45575493/what-does-required-in-openapi-really-mean

**Comment by cbartz** [comment]
> This is arguably a bug. But the root issue is in Pydantic, see [pydantic/pydantic#1270](https://github.com/pydantic/pydantic/issues/1270).

As this is clearly a bug, this should be stated somewhere in the FastAPI docs (with the workaround given by specifying ``nullable=True`` for each optional field).

I've used auto-generated clients that refused to work with a FastAPI server because null values were sent for optional but non-nullable fields.

**Comment by dotmitsu** [comment]
Why this code works normal with Redoc but don't with Swagger?
```
@app.get("/api/v1", tags=["test"])
async def greet() -> None:
    return None
```
In Redoc it shows Response 200 Null
In Swagger it shows Responce 200 "string"
![изображение](https://github.com/tiangolo/fastapi/assets/42657211/a3bc768c-449f-48ff-8505-c35ecb3a7790)
![изображение](https://github.com/tiangolo/fastapi/assets/42657211/a1cccffd-01f3-4a6c-bb61-0a314e2e2b34)

**Comment by ybressler** [comment]
To provide clarification, from the Pydantic side of things: marking a field as `Optional[...]` will allow it to have `None` as a value. However, I imagine `nullable` is different in that it assigns `None` as the default value – whereas pydantic may not. (In V1, pydantic will assign `None` as a default value to an optionally annotated field – but not in V2.)

**Reply by dotmitsu** [reply]
But I didn't want value with Null. I want a response body with null. By default it is null, but swagger shows string in docs

**Reply by ybressler** [reply]
Does marking as `nullable` fix this? If not, perhaps switching from `Optional[str]` to `str | None`? Or `None | str` (switching the order). Shooting in the dark here.

**Reply by dotmitsu** [reply]
No, It doesn't fix this. You can see it if you create empty fast api project and create endpoint which must returns 200 and empty html body.

**Comment by m7142yosuke** [comment]
I noticed that the issue https://github.com/pydantic/pydantic/issues/1270 has been closed. Could you please provide an update on any progress or resolutions that were made in relation to this issue?

**Comment by YuriiMotov** [comment, marked as answer]
This is a [bug in Pydantic V1](https://github.com/pydantic/pydantic/issues/1270) and it will not be fixed.

So, `a: Optional[int] = Field(..., nullable=True)` is the only option.

But better switch to Pydantic V2
